src/aspire/source/__init__.py

```diff
@@ -526,24 +529,174 @@ def vol_forward(self, vol, start, num):
+        )
+        if save_images:
+            logger.info("save images into MRCS file")
+            self._save_images(
+                starfile_filepath,
+                batch_size=batch_size,
+                save_mode=save_mode,
+                overwrite=overwrite,
+            )
+
+    def _save_metadata(self, starfile_filepath, batch_size=512, save_mode=None):
+        """
+        Save updated metadata to a STAR file
+
+        :param starfile_filepath: Path to STAR file where we want to
+            save image_source
+        :param batch_size: Batch size of images to query from the
+            `ImageSource` object. Every `batch_size` rows, entries are
+            written to STAR file.
+        :param save_mode: Whether to save all images in a single or
+            multiple files in batch size.
+        :return: None
+        """
 
-        save_star(
-            self,
-            starfile_filepath,
-            batch_size=batch_size,
-            save_mode=save_mode,
-            overwrite=overwrite,
+        df = self._metadata.copy()
+        # Drop any column that doesn't start with a *single* underscore
+        df = df.drop(
+            [
+                str(col)
+                for col in df.columns
+                if not col.startswith("_") or col.startswith("__")
+            ],
+            axis=1,
         )
 
+        # Create a new column that we will be populating in the loop below
+        df["_rlnImageName"] = ""
+
+        with open(starfile_filepath, "w") as f:
+            if save_mode == "single":
+                # Save all images into one single mrc file
+                fname = os.path.basename(starfile_filepath)
+                fstem = os.path.splitext(fname)[0]
+                mrcs_filename = f"{fstem}_{0}_{self.n-1}.mrcs"
+
+                # Then set name in dataframe for the StarFile
+                df["_rlnImageName"][0 : self.n] = pd.Series(
+                    [f"{j + 1:06}@{mrcs_filename}" for j in range(self.n)]
+                )
+
+            else:
+                # save all images into multiple mrc files in batch size
+                for i_start in np.arange(0, self.n, batch_size):
+                    i_end = min(self.n, i_start + batch_size)
+                    num = i_end - i_start
+                    mrcs_filename = (
+                        os.path.splitext(os.path.basename(starfile_filepath))[0]
+                        + f"_{i_start}_{i_end-1}.mrcs"
+                    )
+
+                    df["_rlnImageName"][i_start:i_end] = pd.Series(
+                        ["{0:06}@{1}".format(j + 1, mrcs_filename) for j in range(num)]
+                    )
+
+            # initial the star file object and save it
+            starfile = StarFile(blocks=[StarFileBlock(loops=[df])])
+            starfile.save(f)
+
+        self._metadata_updated = df
+
+    def _save_images(
+        self, starfile_filepath, batch_size=512, save_mode=None, overwrite=False
```

Comment:
Can't `batch_size` and `save_mode` be inferred from the list of filenames?

Reply:
The `batch_size` can not be inferred from the list of filenames when the `save_mode` is `Single`. But `save_mode` can be obtained from the number of unique file names.  `save_mode` was removed in commit ac65471.